core/src/main/java/com/datastax/oss/driver/internal/core/metadata/DefaultTabletMap.java

```diff
@@ -0,0 +1,288 @@
+      return null;
+    }
+    Tablet result = mapping.get(key).ceiling(DefaultTablet.malformedTablet(token));
+    if (result == null || result.getFirstToken() >= token) {
+      LOG.trace(
+          "Could not find tablet for {}.{} that owns token {}. Returning null.",
+          keyspace,
+          table,
+          token);
+      return null;
+    }
+    return result;
+  }
+
+  @Override
+  public void addTablet(Tablet tablet) {
+    KeyspaceTableNamePair ktPair =
+        new KeyspaceTableNamePair(tablet.getKeyspaceName(), tablet.getTableName());
+
+    // Get existing tablets for given table
+    NavigableSet<Tablet> existingTablets =
+        mapping.computeIfAbsent(ktPair, k -> new ConcurrentSkipListSet<>());
+    // Single tablet token range is represented by (firstToken, lastToken] interval
+    // We need to do two sweeps: remove overlapping tablets by looking at lastToken of existing
+    // tablets
+    // and then by looking at firstToken of existing tablets. Currently, the tablets are sorted
+    // according
+    // to their lastTokens.
+
+    // First sweep: remove all tablets whose lastToken is inside this interval
+    Iterator<Tablet> it = existingTablets.headSet(tablet, true).descendingIterator();
+    while (it.hasNext()) {
+      Tablet nextTablet = it.next();
+      if (nextTablet.getLastToken() <= tablet.getFirstToken()) {
+        break;
+      }
+      it.remove();
+    }
+
+    // Second sweep: remove all tablets whose firstToken is inside this tuple's (firstToken,
+    // lastToken]
+    // After the first sweep, this theoretically should remove at most 1 tablet
+    it = existingTablets.tailSet(tablet, true).iterator();
+    while (it.hasNext()) {
+      Tablet nextTablet = it.next();
+      if (nextTablet.getFirstToken() >= tablet.getLastToken()) {
+        break;
+      }
+      it.remove();
+    }
+
+    // Add new (now) non-overlapping tablet
+    existingTablets.add(tablet);
+  }
+
+  /**
+   * Represents a single tablet created from tablets-routing-v1 custom payload. Its {@code
+   * compareTo} implementation intentionally relies solely on {@code lastToken} in order to allow
+   * for quick lookup on sorted Collections based just on the token value. Its token range is
+   * (firstToken, lastToken], meaning firstToken is not included.
+   */
+  public static class DefaultTablet implements Tablet {
+
+    private static final Logger LOG = LoggerFactory.getLogger(DefaultTablet.class);
+    private final String keyspaceName;
+    private final UUID tableId; // currently null almost everywhere
+    private final String tableName;
+    private final long firstToken;
+    private final long lastToken;
+    private final Set<NodeShardPair> replicas;
+    private final Set<Node> replicaNodes;
+    private final Map<Node, Integer> replicaShards;
+
+    @VisibleForTesting
+    DefaultTablet(
+        String keyspaceName,
+        UUID tableId,
+        String tableName,
```

Comment:
Most of the fields are there because I've been just copying the structure I've seen in design doc.
The only function this has right now is that it simplifies passing it around. I can remove this, but I'll have to pass keyspace and table as separate arguments. This will still save memory because it won't be preserved long term inside the map.

Reply:
Removed those fields